janitor/utils.py

```diff
@@ -321,8 +322,9 @@ def simple_sum(alpha, beta):
     """  # noqa: E501
 
     def decorator(func):
+        @wraps(func)
         def emit_warning(*args, **kwargs):
-            warn(message, FutureWarning)
+            warn(message, FutureWarning, stacklevel=2)
```

Comment:
pandas has a function called `find_stack_level`.
https://github.com/pandas-dev/pandas/blob/1d5ce5b3b42cebd4b49ef3b06b50b78e09b1d91b/pandas/util/_exceptions.py#L30-L51

```python
def find_stack_level() -> int:
    """
    Find the first place in the stack that is not inside pandas
    (tests notwithstanding).
    """

    import pandas as pd

    pkg_dir = os.path.dirname(pd.__file__)
    test_dir = os.path.join(pkg_dir, "tests")

    # https://stackoverflow.com/questions/17407119/python-inspect-stack-is-slow
    frame = inspect.currentframe()
    n = 0
    while frame:
        fname = inspect.getfile(frame)
        if fname.startswith(pkg_dir) and not fname.startswith(test_dir):
            frame = frame.f_back
            n += 1
        else:
            break
    return n
```

Reply:
We use the decorator to wrap function, so a fixed `stacklevel` sometimes can't work well.

Reply:
Sorry @Zeroto521 I don't quite get what you are referring to here

Reply:
```python
# temp.py
from functools import wraps
from typing import Hashable
from warnings import warn

import pandas_flavor as pf
import pandas as pd


def refactored_function(message: str):
    def decorator(func):
        @wraps(func)
        def emit_warning(*args, **kwargs):
            warn(message, FutureWarning, stacklevel=2)
            return func(*args, **kwargs)

        return emit_warning

    return decorator


@pf.register_dataframe_method
@refactored_function("This function is deprecated")
def plus_one(df: pd.DataFrame, column: Hashable) -> pd.DataFrame:
    df = df.copy()
    df[column] += 1

    return df


if __name__ == "__main__":
    df = pd.DataFrame({"a": [0, 1]})
    print(df.plus_one("a"))
```

Because of the decorator, the place of the emitting the warning message is not correct.

```console
$ python .\temp.py
C:\Software\miniforge3\envs\janitor-dev\lib\site-packages\pandas_flavor\register.py:29: FutureWarning: This function is deprecated
  return method(self._obj, *args, **kwargs)
   a
0  1
```

Reply:
Oh I get U now, the stacklevel

Reply:
@Zeroto521 how do you suggest I adapt the `find_stack_level` to the `refactored_function` function?

Reply:
Add this version `find_stack_level` method for `janitor/util.py`.

```python
# from pandas https://github.com/pandas-dev/pandas/blob/1d5ce5b3b42cebd4b49ef3b06b50b78e09b1d91b/pandas/util/_exceptions.py#L30-L51
def find_stack_level() -> int:
    """
    Find the first place in the stack that is not inside pandas
    (tests notwithstanding).
    """

    import janitor as jn

    pkg_dir = os.path.dirname(jn.__file__)
    test_dir = os.path.join(pkg_dir, "tests")

    # https://stackoverflow.com/questions/17407119/python-inspect-stack-is-slow
    frame = inspect.currentframe()
    n = 0
    while frame:
        fname = inspect.getfile(frame)
        if fname.startswith(pkg_dir) and not fname.startswith(test_dir):
            frame = frame.f_back
            n += 1
        else:
            break
    return n
```

then

```python
def refactored_function(
    message: str,
    category: Except = FutureWarning,
    stacklevel: int = find_stack_level(),
):
    def decorator(func):
        @wraps(func)
        def emit_warning(*args, **kwargs):
            warn(message, category=category, stacklevel=stacklevel)
            return func(*args, **kwargs)

        return emit_warning

    return decorator
```

Reply:
janitor folder structures are different from pandas.
These lines should be the followings.

```python
    pkg_dir = os.path.abspath(os.path.dirname(jn.__file__))
    test_dir = os.path.join(os.path.dirname(pkg_dir), "tests")
```